Source/WebCore/dom/ChildrenToAsyncDelete.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2007, 2014, 2015 Apple Inc. All rights reserved.
+ * Copyright (C) 2007 David Smith (catfish.man@gmail.com)
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1.  Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ * 2.  Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ * 3.  Neither the name of Apple Inc. ("Apple") nor the names of
+ *     its contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "config.h"
+#include "ChildrenToAsyncDelete.h"
+
+#include "HTMLElement.h"
+
+namespace WebCore {
+
+static HashSet<AtomString>& allowList()
+{
+    static NeverDestroyed<HashSet<AtomString>> allowList = std::initializer_list<AtomString> {
+    HTMLNames::divTag->localName(), HTMLNames::aTag->localName(),
+    HTMLNames::addressTag->localName(), HTMLNames::bTag->localName(), HTMLNames::bdiTag->localName(),
+    HTMLNames::bdoTag->localName(), HTMLNames::citeTag->localName(), HTMLNames::codeTag->localName(),
+    HTMLNames::delTag->localName(), HTMLNames::dfnTag->localName(), HTMLNames::emTag->localName(),
+    HTMLNames::iTag->localName(), HTMLNames::insTag->localName(), HTMLNames::kbdTag->localName(),
+    HTMLNames::labelTag->localName(), HTMLNames::liTag->localName(), HTMLNames::markTag->localName(),
+    HTMLNames::pTag->localName(), HTMLNames::qTag->localName(), HTMLNames::spanTag->localName(),
+    HTMLNames::buttonTag->localName(), HTMLNames::ulTag->localName(), HTMLNames::brTag->localName(),
+    HTMLNames::inputTag->localName(), HTMLNames::strongTag->localName(), HTMLNames::h1Tag->localName(),
```

Comment:
I bet the most of cost comes from destroying input element's shadow tree.
Maybe we can test that hypothesis by just queuing input element and see if that results in the same perf improvement?